receiver/skywalkingreceiver/factory.go

```diff
@@ -0,0 +1,119 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package skywalkingreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/skywalkingreceiver"
+
+// This file implements factory for skywalking receiver.
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configgrpc"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/config/confignet"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver/receiverhelper"
+)
+
+const (
+	typeStr = "skywalking"
+
+	// Protocol values.
+	protoGRPC = "grpc"
+
+	// Default endpoints to bind to.
+	defaultGRPCBindEndpoint = "0.0.0.0:11800"
+	defaultHTTPBindEndpoint = "0.0.0.0:12800"
+)
+
+// NewFactory creates a new Jaeger receiver factory.
+func NewFactory() component.ReceiverFactory {
+	return receiverhelper.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		receiverhelper.WithTraces(createTracesReceiver))
+}
+
+// CreateDefaultConfig creates the default configuration for Skywalking receiver.
+func createDefaultConfig() config.Receiver {
+	return &Config{
+		ReceiverSettings: config.NewReceiverSettings(config.NewComponentID(typeStr)),
+		Protocols: Protocols{
+			GRPC: &configgrpc.GRPCServerSettings{
+				NetAddr: confignet.NetAddr{
+					Endpoint:  defaultGRPCBindEndpoint,
+					Transport: "tcp",
+				},
+			},
+			HTTP: &confighttp.HTTPServerSettings{
+				Endpoint: defaultHTTPBindEndpoint,
+			},
+		},
+	}
+}
+
+// createTracesReceiver creates a trace receiver based on provided config.
+func createTracesReceiver(
+	_ context.Context,
+	set component.ReceiverCreateSettings,
+	cfg config.Receiver,
+	nextConsumer consumer.Traces,
+) (component.TracesReceiver, error) {
+
+	// Convert settings in the source c to configuration struct
+	// that Skywalking receiver understands.
+	rCfg := cfg.(*Config)
+
+	var err error
+	var c configuration
+	// Set ports
+	if rCfg.Protocols.GRPC != nil {
+		c.CollectorGRPCServerSettings = *rCfg.Protocols.GRPC
+		if c.CollectorGRPCPort, err = extractPortFromEndpoint(rCfg.Protocols.GRPC.NetAddr.Endpoint); err != nil {
+			return nil, fmt.Errorf("unable to extract port for the gRPC endpoint: %w", err)
+		}
+	}
+
+	if rCfg.Protocols.HTTP != nil {
+		c.CollectorHTTPSettings = *rCfg.Protocols.HTTP
+		if c.CollectorHTTPPort, err = extractPortFromEndpoint(rCfg.Protocols.HTTP.Endpoint); err != nil {
+			return nil, fmt.Errorf("unable to extract port for the HTTP endpoint: %w", err)
+		}
+	}
+
+	// Create the receiver.
+	return newSkywalkingReceiver(rCfg.ID(), &c, nextConsumer, set), nil
+}
+
+// extract the port number from string in "address:port" format. If the
+// port number cannot be extracted returns an error.
+func extractPortFromEndpoint(endpoint string) (int, error) {
```

Comment:
This is probably not a topic for this PR, but I noticed this function is copied across several components. Perhaps it should be extracted into some common module?

Reply:
make sense, but I am a beginner for `otel collector `, Would you tell me where or which module it should be？

Reply:
I guess may be `internal/tools/` or `internal/common`?

Reply:
This is more like a general observation, I am wondering what is @jpkrohling opinion on that :)